Document forward AD interaction with grad mode

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72216

Fix https://github.com/pytorch/pytorch/issues/72202

Differential Revision: [D33991687](https://our.internmc.facebook.com/intern/diff/D33991687)

soulitzer committed:
Document forward AD interaction with grad mode

[ghstack-poisoned]

soulitzer committed:
Update on "Document forward AD interaction with grad mode"


Fix https://github.com/pytorch/pytorch/issues/72202

[ghstack-poisoned]

albanD commented on torch/autograd/grad_mode.py:
```diff
@@ -104,6 +104,12 @@ class no_grad(_DecoratorContextManager):
 
+    .. note::
+        Currently this API does not apply to :ref:`forward-mode AD <forward-mode-ad>`
+        (possibly subject to change). As a workaround you can unpack your dual tensors.
```
subject to change?

albanD commented on torch/autograd/grad_mode.py:
```diff
@@ -104,6 +104,12 @@ class no_grad(_DecoratorContextManager):
+        Currently this API does not apply to :ref:`forward-mode AD <forward-mode-ad>`
+        (possibly subject to change). As a workaround you can unpack your dual tensors.
+        If you have a use-case that could use this API, we'd like to hear your
+        `thoughts <https://github.com/pytorch/pytorch/issues/new>`_.
```
Not sure we want that here?

albanD reviewed (commented):
Sorry accepted by mistake.

soulitzer committed:
Update on "Document forward AD interaction with grad mode"


Fix https://github.com/pytorch/pytorch/issues/72202

[ghstack-poisoned]

soulitzer committed:
Update on "Document forward AD interaction with grad mode"


Fix https://github.com/pytorch/pytorch/issues/72202

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33991687).

soulitzer commented on torch/autograd/grad_mode.py:
```diff
@@ -104,6 +104,10 @@ class no_grad(_DecoratorContextManager):
 
+    .. note::
+        Currently this API does not apply to :ref:`forward-mode AD <forward-mode-ad>`
+        (subject to change). As a workaround you can unpack your dual tensors.
```
> Still subject to change? I don't think we will ever change that behavior silently. If we really want to add a new one, it will be via another context manager or by adding new flag to this one.

Ah makes sense. I didn't realize "subject to change" connotes that it changes silently.

soulitzer committed:
Update on "Document forward AD interaction with grad mode"


Fix https://github.com/pytorch/pytorch/issues/72202

Differential Revision: [D33991687](https://our.internmc.facebook.com/intern/diff/D33991687)

[ghstack-poisoned]

soulitzer committed:
Update on "Document forward AD interaction with grad mode"


Fix https://github.com/pytorch/pytorch/issues/72202

Differential Revision: [D33991687](https://our.internmc.facebook.com/intern/diff/D33991687)

[ghstack-poisoned]

albanD commented on torch/autograd/grad_mode.py:
```diff
@@ -104,6 +104,11 @@ class no_grad(_DecoratorContextManager):
 
+    .. note::
+        This API does not apply to :ref:`forward-mode AD <forward-mode-ad>`.
+        If you want to disable forward AD for a computation, you can unpack
```
Looks good!
Do we want to replicate the updated message for enable_grad and set_grad_enabled below?

soulitzer replied:
Hmm, those don't have a workaround right? So I'm thinking we don't need a second line for those.

albanD replied:
Sure!

albanD reviewed (approved):
SGTM

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33991687).
